synapse/rest/client/login.py

```diff
@@ -319,6 +319,12 @@ async def _complete_login(
                 canonical_uid = await self.registration_handler.register_user(
                     localpart=UserID.from_string(user_id).localpart
                 )
+            else: # check if user deactivated
+                deactivated = await self.hs.get_datastore().get_user_deactivated_status(canonical_uid)
+                if deactivated:
+                    raise UserDeactivatedError("This account has been deactivated") 
```

Comment:
This doesn't look quite right since it is in the `create_non_existent_users` block.

I'd be curious where we check for deactivation for password users / SSO users and why that isn't being applied properly. Can we consolidate that logic into a single spot?

Reply:
Annoyingly the place where we currently check for user deactivation is while authenticating against the local password database: https://github.com/matrix-org/synapse/blob/e24ff8ebe3d4119d377355402245947f7de61c00/synapse/handlers/auth.py#L1379-L1383.

Thankfully we can check if a user is deactivated via the `deactivated` column on the `users` table (rather than using an empty password hash as a heuristic).

So I'd recommend pulling the user deactivated check out of `_check_local_password` and instead moving it into `_validate_userid_login`, such that it can be applied to all user logins; not just those against the local password database.

Reply:
Thanks for digging into that @anoadragon453! That sounds sensible to me... I think we did something like this for SSO too at some point.

Looks like we did something specific for SSO too (see #7240).

Reply:
Ah, it makes sense that you'd want to show an error page for SSO flows, rather than raising an `UserDeactivatedError`.

Still, the local password and JWT flows could be consolidated.